apis/python/src/tiledbsoma/dataframe.py

```diff
@@ -67,105 +50,34 @@ def create_legacy(
-                domain = None, None
-            elif np.issubdtype(dtype, NPInteger):
-                iinfo = np.iinfo(cast(NPInteger, dtype))
-                domain = iinfo.min, iinfo.max - 1
-            elif np.issubdtype(dtype, NPFloating):
-                finfo = np.finfo(cast(NPFloating, dtype))
-                domain = finfo.min, finfo.max
-            else:
-                raise TypeError(f"Unsupported dtype {dtype}")
-
-            # Default 2048 mods to 0 for 8-bit types and 0 is an invalid extent
-            extent = tiledb_create_options.dim_tile(index_column_name)
-            if isinstance(dtype, np.dtype) and dtype.itemsize == 1:
-                extent = 64
-
-            dim = tiledb.Dim(
-                name=index_column_name,
-                domain=domain,
-                tile=extent,
-                dtype=dtype,
-                filters=tiledb_create_options.dim_filters(
-                    index_column_name,
-                    [
-                        dict(
-                            _type="ZstdFilter",
-                            level=tiledb_create_options.string_dim_zstd_level(),
-                        )
-                    ],
-                ),
-            )
-            dims.append(dim)
-
-        dom = tiledb.Domain(dims, ctx=self._ctx)
-
-        attrs = []
-        for attr_name in schema.names:
-            if attr_name in index_column_names:
-                continue
-            attr = tiledb.Attr(
-                name=attr_name,
-                dtype=util_arrow.tiledb_type_from_arrow_type(
-                    schema.field(attr_name).type
-                ),
-                filters=tiledb_create_options.attr_filters(attr_name, ["ZstdFilter"]),
-                ctx=self._ctx,
-            )
-            attrs.append(attr)
-
-        cell_order, tile_order = tiledb_create_options.cell_tile_orders()
-
-        sch = tiledb.ArraySchema(
-            domain=dom,
-            attrs=attrs,
-            sparse=True,
-            allows_duplicates=False,
-            offsets_filters=tiledb_create_options.offsets_filters(),
-            capacity=tiledb_create_options.get("capacity", 100000),
-            cell_order=cell_order,
-            # As of TileDB core 2.8.2, we cannot consolidate string-indexed sparse arrays with
-            # col-major tile order: so we write ``X`` with row-major tile order.
-            tile_order=tile_order,
-            ctx=self._ctx,
+        handle = cls._create_internal(uri, tdb_schema, context)
+        return cls(
+            handle,
+            _index_column_names=tuple(index_column_names),
+            _this_is_internal_only="tiledbsoma-internal-code",
         )
-        self._is_sparse = sch.sparse
 
-        tiledb.Array.create(self._uri, sch)
+    def __init__(
+        self,
+        handle: tiledb.Array,
+        *,
+        # Hints to pre-fill cache entries.
+        _index_column_names: Optional[Tuple[str, ...]] = None,
```

Comment:
I'm not sure caching the index column names makes any sense given the change to stateful handles.  I believe that it was originally put in place to avoid repeated calls to `open` on the array (just to get the dimension names).  With an open handle guaranteed, getting dim names is cheap, ergo caching is not worth the complexity.

Reply:
I’ll leave this as-is for now to reduce churn here.